library/core/src/time.rs

```diff
@@ -732,52 +746,113 @@ impl Duration {
     ///
     /// # Examples
     /// ```
     /// #![feature(duration_checked_float)]
     ///
     /// use std::time::Duration;
     ///
-    /// let dur = Duration::try_from_secs_f64(2.7);
-    /// assert_eq!(dur, Ok(Duration::new(2, 700_000_000)));
-    ///
-    /// let negative = Duration::try_from_secs_f64(-5.0);
-    /// assert!(negative.is_err());
+    /// let res = Duration::try_from_secs_f64(0.0);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f64(1e-20);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f64(4.2e-7);
+    /// assert_eq!(res, Ok(Duration::new(0, 420)));
+    /// let res = Duration::try_from_secs_f64(2.7);
+    /// assert_eq!(res, Ok(Duration::new(2, 700_000_000)));
+    /// let res = Duration::try_from_secs_f64(3e10);
+    /// assert_eq!(res, Ok(Duration::new(30_000_000_000, 0)));
+    /// // subnormal float
+    /// let res = Duration::try_from_secs_f64(f64::from_bits(1));
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// // conversion uses truncation, not rounding
+    /// let res = Duration::try_from_secs_f32(0.999e-9);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    ///
+    /// let res = Duration::try_from_secs_f64(-5.0);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f64(f64::NAN);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f64(2e19);
+    /// assert!(res.is_err());
     /// ```
     #[unstable(feature = "duration_checked_float", issue = "83400")]
     #[inline]
-    pub const fn try_from_secs_f64(secs: f64) -> Result<Duration, FromSecsError> {
-        const MAX_NANOS_F64: f64 = ((u64::MAX as u128 + 1) * (NANOS_PER_SEC as u128)) as f64;
-        let nanos = secs * (NANOS_PER_SEC as f64);
-        if !nanos.is_finite() {
-            Err(FromSecsError { kind: FromSecsErrorKind::NonFinite })
-        } else if nanos >= MAX_NANOS_F64 {
-            Err(FromSecsError { kind: FromSecsErrorKind::Overflow })
-        } else if nanos < 0.0 {
-            Err(FromSecsError { kind: FromSecsErrorKind::Negative })
-        } else {
-            let nanos = nanos as u128;
-            Ok(Duration {
-                secs: (nanos / (NANOS_PER_SEC as u128)) as u64,
-                nanos: (nanos % (NANOS_PER_SEC as u128)) as u32,
-            })
+    pub const fn try_from_secs_f64(secs: f64) -> Result<Duration, FromFloatSecsError> {
+        const FRACT_BITS: u64 = 52;
+        const FRACT_MASK: u64 = (1 << FRACT_BITS) - 1;
+        const EXP_BITS: u64 = 11;
+        const EXP_MASK: u64 = (1 << EXP_BITS) - 1;
+        const OFFSET: u64 = 32;
+        const DIV1: u128 = 1u128 << FRACT_BITS;
+        const DIV2: u128 = DIV1 << OFFSET;
+
+        if secs.is_sign_negative() {
+            return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::Negative });
         }
+
+        let bits = secs.to_bits();
```

Comment:
I prototyped an approach that uses macros and this is what I came up with:

<details>
<summary>Diff</summary>

```diff
diff --git a/library/core/src/lib.rs b/library/core/src/lib.rs
index 63c42068aad..6667dff5b5c 100644
--- a/library/core/src/lib.rs
+++ b/library/core/src/lib.rs
@@ -191,6 +191,8 @@
 #![feature(try_blocks)]
 #![feature(unboxed_closures)]
 #![feature(unsized_fn_params)]
+#![feature(exclusive_range_pattern)]
+#![feature(const_convert)]
 //
 // Target features:
 #![feature(aarch64_target_feature)]
diff --git a/library/core/src/time.rs b/library/core/src/time.rs
index 968c94b3cf3..7ab1d6feb7a 100644
--- a/library/core/src/time.rs
+++ b/library/core/src/time.rs
@@ -742,85 +742,6 @@ pub const fn from_secs_f64(secs: f64) -> Duration {
         }
     }
 
-    /// The checked version of [`from_secs_f64`].
-    ///
-    /// [`from_secs_f64`]: Duration::from_secs_f64
-    ///
-    /// This constructor will return an `Err` if `secs` is negative, overflows `Duration` or not finite.
-    ///
-    /// # Examples
-    /// ```
-    /// #![feature(duration_checked_float)]
-    ///
-    /// use std::time::Duration;
-    ///
-    /// let res = Duration::try_from_secs_f64(0.0);
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    /// let res = Duration::try_from_secs_f64(1e-20);
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    /// let res = Duration::try_from_secs_f64(4.2e-7);
-    /// assert_eq!(res, Ok(Duration::new(0, 420)));
-    /// let res = Duration::try_from_secs_f64(2.7);
-    /// assert_eq!(res, Ok(Duration::new(2, 700_000_000)));
-    /// let res = Duration::try_from_secs_f64(3e10);
-    /// assert_eq!(res, Ok(Duration::new(30_000_000_000, 0)));
-    /// // subnormal float
-    /// let res = Duration::try_from_secs_f64(f64::from_bits(1));
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    /// // conversion uses truncation, not rounding
-    /// let res = Duration::try_from_secs_f32(0.999e-9);
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    ///
-    /// let res = Duration::try_from_secs_f64(-5.0);
-    /// assert!(res.is_err());
-    /// let res = Duration::try_from_secs_f64(f64::NAN);
-    /// assert!(res.is_err());
-    /// let res = Duration::try_from_secs_f64(2e19);
-    /// assert!(res.is_err());
-    /// ```
-    #[unstable(feature = "duration_checked_float", issue = "83400")]
-    #[inline]
-    pub const fn try_from_secs_f64(secs: f64) -> Result<Duration, FromFloatSecsError> {
-        const MANT_BITS: i16 = 52;
-        const MANT_MASK: u64 = (1 << MANT_BITS) - 1;
-        const EXP_BITS: i16 = 11;
-        const EXP_MASK: u64 = (1 << EXP_BITS) - 1;
-        const OFFSET: i16 = 44;
-        const DIV1: u128 = 1u128 << MANT_BITS;
-        const DIV2: u128 = 1u128 << (MANT_BITS + OFFSET);
-
-        if secs.is_sign_negative() {
-            return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::Negative });
-        }
-
-        let bits = secs.to_bits();
-        let mant = (bits & MANT_MASK) | (MANT_MASK + 1);
-        let exp = ((bits >> MANT_BITS) & EXP_MASK) as i16 - 1023;
-
-        // note: max `f64` for `exp` equal to -31 is 9.31e-10, which is less than 1 ns
-        let (secs, nanos) = match exp {
-            -1023..=-31 => (0, 0),
-            -30..=-1 => {
-                let t = u128::from(mant) << (OFFSET + exp);
-                let nanos = (u128::from(NANOS_PER_SEC) * t) / DIV2;
-                (0, nanos as u32)
-            }
-            0..=51 => {
-                let secs = mant >> (MANT_BITS - exp);
-                let t = (mant << exp) & MANT_MASK;
-                let nanos = (u128::from(NANOS_PER_SEC) * u128::from(t)) / DIV1;
-                (secs, nanos as u32)
-            }
-            52..=63 => {
-                let secs = mant << (exp - MANT_BITS);
-                (secs, 0)
-            }
-            _ => return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::OverflowOrNan }),
-        };
-
-        Ok(Duration { secs, nanos })
-    }
-
     /// Creates a new `Duration` from the specified number of seconds represented
     /// as `f32`.
     ///
@@ -859,85 +780,6 @@ pub const fn from_secs_f32(secs: f32) -> Duration {
         }
     }
 
-    /// The checked version of [`from_secs_f32`].
-    ///
-    /// [`from_secs_f32`]: Duration::from_secs_f32
-    ///
-    /// This constructor will return an `Err` if `secs` is negative, overflows `Duration` or not finite.
-    ///
-    /// # Examples
-    /// ```
-    /// #![feature(duration_checked_float)]
-    ///
-    /// use std::time::Duration;
-    ///
-    /// let res = Duration::try_from_secs_f32(0.0);
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    /// let res = Duration::try_from_secs_f32(1e-20);
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    /// let res = Duration::try_from_secs_f32(4.2e-7);
-    /// assert_eq!(res, Ok(Duration::new(0, 419)));
-    /// let res = Duration::try_from_secs_f32(2.7);
-    /// assert_eq!(res, Ok(Duration::new(2, 700_000_047)));
-    /// let res = Duration::try_from_secs_f32(3e10);
-    /// assert_eq!(res, Ok(Duration::new(30_000_001_024, 0)));
-    /// // subnormal float:
-    /// let res = Duration::try_from_secs_f32(f32::from_bits(1));
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    /// // conversion uses truncation, not rounding
-    /// let res = Duration::try_from_secs_f32(0.999e-9);
-    /// assert_eq!(res, Ok(Duration::new(0, 0)));
-    ///
-    /// let res = Duration::try_from_secs_f32(-5.0);
-    /// assert!(res.is_err());
-    /// let res = Duration::try_from_secs_f32(f32::NAN);
-    /// assert!(res.is_err());
-    /// let res = Duration::try_from_secs_f32(2e19);
-    /// assert!(res.is_err());
-    /// ```
-    #[unstable(feature = "duration_checked_float", issue = "83400")]
-    #[inline]
-    pub const fn try_from_secs_f32(secs: f32) -> Result<Duration, FromFloatSecsError> {
-        const MANT_BITS: i16 = 23;
-        const MANT_MASK: u32 = (1 << MANT_BITS) - 1;
-        const EXP_BITS: i16 = 8;
-        const EXP_MASK: u32 = (1 << EXP_BITS) - 1;
-        const OFFSET: i16 = 41;
-        const DIV1: u64 = 1u64 << MANT_BITS;
-        const DIV2: u128 = 1u128 << (MANT_BITS + OFFSET);
-
-        if secs.is_sign_negative() {
-            return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::Negative });
-        }
-
-        let bits = secs.to_bits();
-        let mant = (bits & MANT_MASK) | (MANT_MASK + 1);
-        let exp = ((bits >> MANT_BITS) & EXP_MASK) as i16 - 127;
-
-        // note: max `f32` for `exp` equal to -31 is 9.31e-10, which is less than 1 ns
-        let (secs, nanos) = match exp {
-            -127..=-31 => (0, 0),
-            -30..=-1 => {
-                let t = u64::from(mant) << (OFFSET + exp);
-                let nanos = (u128::from(NANOS_PER_SEC) * u128::from(t)) / DIV2;
-                (0, nanos as u32)
-            }
-            0..=22 => {
-                let secs = mant >> (MANT_BITS - exp);
-                let t = (mant << exp) & MANT_MASK;
-                let nanos = (u64::from(NANOS_PER_SEC) * u64::from(t)) / DIV1;
-                (u64::from(secs), nanos as u32)
-            }
-            23..=63 => {
-                let secs = u64::from(mant) << (exp - MANT_BITS);
-                (secs, 0)
-            }
-            _ => return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::OverflowOrNan }),
-        };
-
-        Ok(Duration { secs, nanos })
-    }
-
     /// Multiplies `Duration` by `f64`.
     ///
     /// # Panics
@@ -1071,6 +913,149 @@ pub const fn div_duration_f32(self, rhs: Duration) -> f32 {
     }
 }
 
+macro_rules! try_from_secs {
+    (
+        secs = $secs: expr,
+        mantissa_bits = $mant_bits: literal,
+        exponent_bits = $exp_bits: literal,
+        offset = $offset: literal,
+        min_exp = $min_exp: literal
+    ) => {{
+        if $secs.is_sign_negative() {
+            return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::Negative });
+        }
+
+        let bits = $secs.to_bits();
+        let mant_mask = ((1 << $mant_bits) - 1);
+        let mant = (bits & mant_mask) | (mant_mask + 1);
+        let exp_mask = (1 << $exp_bits) - 1;
+        let exp = ((bits >> $mant_bits) & exp_mask) as i16 + $min_exp;
+
+        let (secs, nanos) = match exp {
+            // the input represents less than 1ns.
+            $min_exp..-30 => (0u64, 0u32),
+            // the input is less than 1 second
+            -30..0 => {
+                let t = u128::from(mant) << ($offset + exp);
+                let nanos = (u128::from(NANOS_PER_SEC) * t) >> ($mant_bits + $offset);
+                (0, nanos as u32)
+            }
+            0..$mant_bits => {
+                let secs = mant >> ($mant_bits as i16 - exp);
+                let t = (mant << exp) & mant_mask;
+                let nanos = (u128::from(NANOS_PER_SEC) * u128::from(t)) >> $mant_bits;
+                (u64::from(secs), nanos as u32)
+            }
+            // the input has no fractional part
+            $mant_bits..64 => {
+                let secs = u64::from(mant) << (exp - $mant_bits as i16);
+                (secs, 0)
+            }
+            _ => return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::OverflowOrNan }),
+        };
+
+        Ok(Duration { secs, nanos })
+    }};
+}
+
+impl Duration {
+    /// The checked version of [`from_secs_f64`].
+    ///
+    /// [`from_secs_f64`]: Duration::from_secs_f64
+    ///
+    /// This constructor will return an `Err` if `secs` is negative, overflows `Duration` or not finite.
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(duration_checked_float)]
+    ///
+    /// use std::time::Duration;
+    ///
+    /// let res = Duration::try_from_secs_f64(0.0);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f64(1e-20);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f64(4.2e-7);
+    /// assert_eq!(res, Ok(Duration::new(0, 420)));
+    /// let res = Duration::try_from_secs_f64(2.7);
+    /// assert_eq!(res, Ok(Duration::new(2, 700_000_000)));
+    /// let res = Duration::try_from_secs_f64(3e10);
+    /// assert_eq!(res, Ok(Duration::new(30_000_000_000, 0)));
+    /// // subnormal float
+    /// let res = Duration::try_from_secs_f64(f64::from_bits(1));
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// // conversion uses truncation, not rounding
+    /// let res = Duration::try_from_secs_f32(0.999e-9);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    ///
+    /// let res = Duration::try_from_secs_f64(-5.0);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f64(f64::NAN);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f64(2e19);
+    /// assert!(res.is_err());
+    /// ```
+    #[unstable(feature = "duration_checked_float", issue = "83400")]
+    #[inline]
+    pub const fn try_from_secs_f64(secs: f64) -> Result<Duration, FromFloatSecsError> {
+        try_from_secs!(
+            secs = secs,
+            mantissa_bits = 52,
+            exponent_bits = 11,
+            offset = 44,
+            min_exp = -1023
+        )
+    }
+
+    /// The checked version of [`from_secs_f32`].
+    ///
+    /// [`from_secs_f32`]: Duration::from_secs_f32
+    ///
+    /// This constructor will return an `Err` if `secs` is negative, overflows `Duration` or not finite.
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(duration_checked_float)]
+    ///
+    /// use std::time::Duration;
+    ///
+    /// let res = Duration::try_from_secs_f32(0.0);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f32(1e-20);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f32(4.2e-7);
+    /// assert_eq!(res, Ok(Duration::new(0, 419)));
+    /// let res = Duration::try_from_secs_f32(2.7);
+    /// assert_eq!(res, Ok(Duration::new(2, 700_000_047)));
+    /// let res = Duration::try_from_secs_f32(3e10);
+    /// assert_eq!(res, Ok(Duration::new(30_000_001_024, 0)));
+    /// // subnormal float:
+    /// let res = Duration::try_from_secs_f32(f32::from_bits(1));
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// // conversion uses truncation, not rounding
+    /// let res = Duration::try_from_secs_f32(0.999e-9);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    ///
+    /// let res = Duration::try_from_secs_f32(-5.0);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f32(f32::NAN);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f32(2e19);
+    /// assert!(res.is_err());
+    /// ```
+    #[unstable(feature = "duration_checked_float", issue = "83400")]
+    #[inline]
+    pub const fn try_from_secs_f32(secs: f32) -> Result<Duration, FromFloatSecsError> {
+        try_from_secs!(
+            secs = secs,
+            mantissa_bits = 23,
+            exponent_bits = 8,
+            offset = 41,
+            min_exp = -127
+        )
+    }
+}
+
 #[stable(feature = "duration", since = "1.3.0")]
 impl Add for Duration {
     type Output = Duration;
```

</details>

Despite the fact that this is using a macro, I still find it somewhat more readable now – a macro forced me to use fewer magic numbers, and ultimately there's just one implementation now, which I feel is a significant boost to maintainability still (even though the number of lines didn't change significantly at all)

What do you think?

(NB: I didn't actually rigorously test this implementation)

Reply:
I am fine with using a macro, if you think it's a proffered approach. After a cursory look I have 3 nits:
- In `try_from_secs_f32` in the third arm you use wider operation than strictly necessary. Compiler may optimize it, but I would prefer to explicitly use `u64` instead of `u128`.
- A bit too much reliance on type inference for my taste.
- `min_exp` can be computed as `1 - (1i16 << (exponent_bits - 1))`, so it can be removed.

I will try push variation of your macro a bit later.